File: backend/app/schemas/workflow.py
from datetime import datetime
from typing import Any, Dict, List, Optional

from pydantic import BaseModel


# WorkflowType schemas
class WorkflowTypeBase(BaseModel):
    name: str
    description: Optional[str] = None
    icon: Optional[str] = None
    color: Optional[str] = None
    is_active: bool = True


class WorkflowTypeCreate(WorkflowTypeBase):
    pass


class WorkflowTypeUpdate(BaseModel):
    name: Optional[str] = None
    description: Optional[str] = None
    icon: Optional[str] = None
    color: Optional[str] = None
    is_active: Optional[bool] = None


class WorkflowType(WorkflowTypeBase):
    id: int
    created_by: int
    created_at: datetime
    updated_at: Optional[datetime] = None

    class Config:
        from_attributes = True


# WorkflowState schemas
class WorkflowStateBase(BaseModel):
    name: str
    label: str
    description: Optional[str] = None
    color: str = "#6B7280"
    bg_color: str = "#F9FAFB"
    icon: str = "circle"
    is_initial: bool = False
    is_final: bool = False


class WorkflowStateCreate(WorkflowStateBase):
    pass


class WorkflowStateUpdate(BaseModel):
    name: Optional[str] = None
    label: Optional[str] = None
    description: Optional[str] = None
    color: Optional[str] = None
    bg_color: Optional[str] = None
    icon: Optional[str] = None
    is_initial: Optional[bool] = None
    is_final: Optional[bool] = None


class WorkflowState(WorkflowStateBase):
    id: int
    created_at: datetime

    class Config:
        from_attributes = True


# WorkflowTemplate schemas
class WorkflowTemplateBase(BaseModel):
    name: str
    description: Optional[str] = None
    workflow_type_id: int
    default_state_id: Optional[int] = None
    is_active: bool = True
    nodes: List[Dict[str, Any]] = []
    edges: List[Dict[str, Any]] = []
    settings: Dict[str, Any] = {}


class WorkflowTemplateCreate(WorkflowTemplateBase):
    pass


class WorkflowTemplateUpdate(BaseModel):
    name: Optional[str] = None
    description: Optional[str] = None
    default_state_id: Optional[int] = None
    is_active: Optional[bool] = None
    nodes: Optional[List[Dict[str, Any]]] = None
    edges: Optional[List[Dict[str, Any]]] = None
    settings: Optional[Dict[str, Any]] = None


class WorkflowTemplate(WorkflowTemplateBase):
    id: int
    created_by: int
    created_at: datetime
    updated_at: Optional[datetime] = None
    workflow_type: Optional[WorkflowType] = None
    default_state: Optional[WorkflowState] = None

    class Config:
        from_attributes = True


# WorkflowInstance schemas
class WorkflowInstanceBase(BaseModel):
    template_id: int
    workflow_type_id: int
    entity_id: str
    entity_type: str
    title: Optional[str] = None
    description: Optional[str] = None
    data: Dict[str, Any] = {}


class WorkflowInstanceCreate(WorkflowInstanceBase):
    current_state_id: Optional[int] = None  # Will use template default if not provided


class WorkflowInstanceUpdate(BaseModel):
    current_state_id: Optional[int] = None
    title: Optional[str] = None
    description: Optional[str] = None
    data: Optional[Dict[str, Any]] = None


class WorkflowInstance(WorkflowInstanceBase):
    id: int
    current_state_id: int
    created_by: int
    created_at: datetime
    updated_at: Optional[datetime] = None
    workflow_type: Optional[WorkflowType] = None
    current_state: Optional[WorkflowState] = None
    template: Optional[WorkflowTemplate] = None

    class Config:
        from_attributes = True


# WorkflowHistory schemas
class WorkflowHistoryBase(BaseModel):
    action: str
    comment: Optional[str] = None
    metadata: Dict[str, Any] = {}


class WorkflowHistoryCreate(WorkflowHistoryBase):
    instance_id: int
    from_state_id: Optional[int] = None
    to_state_id: int


class WorkflowHistory(WorkflowHistoryBase):
    id: int
    instance_id: int
    from_state_id: Optional[int] = None
    to_state_id: int
    user_id: int
    timestamp: datetime
    from_state: Optional[WorkflowState] = None
    to_state: Optional[WorkflowState] = None

    class Config:
        from_attributes = True


# WorkflowTransition schemas
class WorkflowTransitionBase(BaseModel):
    template_id: int
    from_state_id: int
    to_state_id: int
    action: str
    label: str
    condition: Optional[str] = None
    requires_approval: bool = False
    allowed_roles: List[str] = []


class WorkflowTransitionCreate(WorkflowTransitionBase):
    pass


class WorkflowTransition(WorkflowTransitionBase):
    id: int
    created_at: datetime
    from_state: Optional[WorkflowState] = None
    to_state: Optional[WorkflowState] = None

    class Config:
        from_attributes = True


# Action schemas for state transitions
class WorkflowAction(BaseModel):
    action: str
    comment: Optional[str] = None
    metadata: Dict[str, Any] = {}

File: frontend/src/shared/services/api/workflow.ts
import { getApiUrl } from "./config";

export interface WorkflowType {
  id: number;
  name: string;
  description?: string;
  icon?: string;
  color?: string;
  is_active: boolean;
  created_by: number;
  created_at: string;
  updated_at?: string;
}

export interface WorkflowTypeCreate {
  name: string;
  description?: string;
  icon?: string;
  color?: string;
  is_active?: boolean;
}

export interface WorkflowTypeUpdate {
  name?: string;
  description?: string;
  icon?: string;
  color?: string;
  is_active?: boolean;
}

export interface WorkflowState {
  id: number;
  name: string;
  label: string;
  description?: string;
  color: string;
  bg_color: string;
  icon: string;
  is_initial: boolean;
  is_final: boolean;
  created_at: string;
}

export interface WorkflowStateCreate {
  name: string;
  label: string;
  description?: string;
  color?: string;
  bg_color?: string;
  icon?: string;
  is_initial?: boolean;
  is_final?: boolean;
}

export interface WorkflowTemplate {
  id: number;
  name: string;
  description?: string;
  workflow_type_id: number;
  default_state_id?: number;
  is_active: boolean;
  nodes: any[];
  edges: any[];
  settings: Record<string, any>;
  created_by: number;
  created_at: string;
  updated_at?: string;
  workflow_type?: WorkflowType;
  default_state?: WorkflowState;
}

export interface WorkflowTemplateCreate {
  name: string;
  description?: string;
  workflow_type_id: number;
  default_state_id?: number;
  is_active?: boolean;
  nodes?: any[];
  edges?: any[];
  settings?: Record<string, any>;
}

export interface WorkflowTemplateUpdate {
  name?: string;
  description?: string;
  default_state_id?: number;
  is_active?: boolean;
  nodes?: any[];
  edges?: any[];
  settings?: Record<string, any>;
}

export interface PaginatedResponse<T> {
  items: T[];
  total: number;
  page: number;
  pages: number;
  size: number;
}

class WorkflowAPI {
  private getAuthHeaders() {
    const token =
      typeof window !== "undefined"
        ? localStorage.getItem("access_token")
        : null;
    return {
      "Content-Type": "application/json",
      ...(token && { Authorization: `Bearer ${token}` }),
    };
  }

  // Workflow Types API
  async getWorkflowTypes(
    params: { skip?: number; limit?: number; search?: string } = {},
  ): Promise<PaginatedResponse<WorkflowType>> {
    const searchParams = new URLSearchParams();
    if (params.skip) searchParams.set("skip", params.skip.toString());
    if (params.limit) searchParams.set("limit", params.limit.toString());
    if (params.search) searchParams.set("search", params.search);

    const response = await fetch(
      `${getApiUrl("/api/v1/workflow-types")}?${searchParams}`,
      {
        headers: this.getAuthHeaders(),
      },
    );

    if (!response.ok) {
      throw new Error("Failed to fetch workflow types");
    }

    return response.json();
  }

  async createWorkflowType(data: WorkflowTypeCreate): Promise<WorkflowType> {
    const response = await fetch(getApiUrl("/api/v1/workflow-types"), {
      method: "POST",
      headers: this.getAuthHeaders(),
      body: JSON.stringify(data),
    });

    if (!response.ok) {
      throw new Error("Failed to create workflow type");
    }

    return response.json();
  }

  async updateWorkflowType(
    id: number,
    data: WorkflowTypeUpdate,
  ): Promise<WorkflowType> {
    const response = await fetch(getApiUrl(`/api/v1/workflow-types/${id}`), {
      method: "PUT",
      headers: this.getAuthHeaders(),
      body: JSON.stringify(data),
    });

    if (!response.ok) {
      throw new Error("Failed to update workflow type");
    }

    return response.json();
  }

  async deleteWorkflowType(id: number): Promise<{ message: string }> {
    const response = await fetch(getApiUrl(`/api/v1/workflow-types/${id}`), {
      method: "DELETE",
      headers: this.getAuthHeaders(),
    });

    if (!response.ok) {
      throw new Error("Failed to delete workflow type");
    }

    return response.json();
  }

  // Workflow States API
  async getWorkflowStates(
    params: { skip?: number; limit?: number; search?: string } = {},
  ): Promise<PaginatedResponse<WorkflowState>> {
    const searchParams = new URLSearchParams();
    if (params.skip) searchParams.set("skip", params.skip.toString());
    if (params.limit) searchParams.set("limit", params.limit.toString());
    if (params.search) searchParams.set("search", params.search);

    const response = await fetch(
      `${getApiUrl("/api/v1/workflow-states")}?${searchParams}`,
      {
        headers: this.getAuthHeaders(),
      },
    );

    if (!response.ok) {
      throw new Error("Failed to fetch workflow states");
    }

    return response.json();
  }

  async createWorkflowState(data: WorkflowStateCreate): Promise<WorkflowState> {
    const response = await fetch(getApiUrl("/api/v1/workflow-states"), {
      method: "POST",
      headers: this.getAuthHeaders(),
      body: JSON.stringify(data),
    });

    if (!response.ok) {
      throw new Error("Failed to create workflow state");
    }

    return response.json();
  }

  // Workflow Templates API
  async getWorkflowTemplates(
    params: {
      skip?: number;
      limit?: number;
      search?: string;
      workflow_type_id?: number;
    } = {},
  ): Promise<PaginatedResponse<WorkflowTemplate>> {
    const searchParams = new URLSearchParams();
    if (params.skip) searchParams.set("skip", params.skip.toString());
    if (params.limit) searchParams.set("limit", params.limit.toString());
    if (params.search) searchParams.set("search", params.search);
    if (params.workflow_type_id)
      searchParams.set("workflow_type_id", params.workflow_type_id.toString());

    const response = await fetch(
      `${getApiUrl("/api/v1/workflow-templates")}?${searchParams}`,
      {
        headers: this.getAuthHeaders(),
      },
    );

    if (!response.ok) {
      throw new Error("Failed to fetch workflow templates");
    }

    return response.json();
  }

  async getWorkflowTemplate(id: number): Promise<WorkflowTemplate> {
    const response = await fetch(
      getApiUrl(`/api/v1/workflow-templates/${id}`),
      {
        headers: this.getAuthHeaders(),
      },
    );

    if (!response.ok) {
      throw new Error("Failed to fetch workflow template");
    }

    return response.json();
  }

  async createWorkflowTemplate(
    data: WorkflowTemplateCreate,
  ): Promise<WorkflowTemplate> {
    const response = await fetch(getApiUrl("/api/v1/workflow-templates"), {
      method: "POST",
      headers: this.getAuthHeaders(),
      body: JSON.stringify(data),
    });

    if (!response.ok) {
      throw new Error("Failed to create workflow template");
    }

    return response.json();
  }

  async updateWorkflowTemplate(
    id: number,
    data: WorkflowTemplateUpdate,
  ): Promise<WorkflowTemplate> {
    const response = await fetch(
      getApiUrl(`/api/v1/workflow-templates/${id}`),
      {
        method: "PUT",
        headers: this.getAuthHeaders(),
        body: JSON.stringify(data),
      },
    );

    if (!response.ok) {
      throw new Error("Failed to update workflow template");
    }

    return response.json();
  }

  async deleteWorkflowTemplate(id: number): Promise<{ message: string }> {
    const response = await fetch(
      getApiUrl(`/api/v1/workflow-templates/${id}`),
      {
        method: "DELETE",
        headers: this.getAuthHeaders(),
      },
    );

    if (!response.ok) {
      throw new Error("Failed to delete workflow template");
    }

    return response.json();
  }

  // Workflow Instances API
  async getWorkflowInstances(
    params: {
      skip?: number;
      limit?: number;
      entity_type?: string;
      status?: string;
      workflow_type_id?: number;
    } = {},
  ): Promise<PaginatedResponse<any>> {
    const searchParams = new URLSearchParams();
    if (params.skip) searchParams.set("skip", params.skip.toString());
    if (params.limit) searchParams.set("limit", params.limit.toString());
    if (params.entity_type) searchParams.set("entity_type", params.entity_type);
    if (params.status) searchParams.set("status", params.status);
    if (params.workflow_type_id)
      searchParams.set("workflow_type_id", params.workflow_type_id.toString());

    const response = await fetch(
      `${getApiUrl("/api/v1/workflow-instances")}?${searchParams}`,
      {
        headers: this.getAuthHeaders(),
      },
    );

    if (!response.ok) {
      throw new Error("Failed to fetch workflow instances");
    }

    return response.json();
  }

  async getWorkflowInstance(id: number): Promise<any> {
    const response = await fetch(
      getApiUrl(`/api/v1/workflow-instances/${id}`),
      {
        headers: this.getAuthHeaders(),
      },
    );

    if (!response.ok) {
      throw new Error("Failed to fetch workflow instance");
    }

    return response.json();
  }

  async createWorkflowInstance(data: any): Promise<any> {
    const response = await fetch(getApiUrl("/api/v1/workflow-instances"), {
      method: "POST",
      headers: this.getAuthHeaders(),
      body: JSON.stringify(data),
    });

    if (!response.ok) {
      throw new Error("Failed to create workflow instance");
    }

    return response.json();
  }

  async updateWorkflowInstance(id: number, data: any): Promise<any> {
    const response = await fetch(
      getApiUrl(`/api/v1/workflow-instances/${id}`),
      {
        method: "PUT",
        headers: this.getAuthHeaders(),
        body: JSON.stringify(data),
      },
    );

    if (!response.ok) {
      throw new Error("Failed to update workflow instance");
    }

    return response.json();
  }

  async executeWorkflowAction(instanceId: number, action: any): Promise<any> {
    const response = await fetch(
      getApiUrl(`/api/v1/workflow-instances/${instanceId}/actions`),
      {
        method: "POST",
        headers: this.getAuthHeaders(),
        body: JSON.stringify(action),
      },
    );

    if (!response.ok) {
      throw new Error("Failed to execute workflow action");
    }

    return response.json();
  }

  async getWorkflowHistory(instanceId: number): Promise<any[]> {
    const response = await fetch(
      getApiUrl(`/api/v1/workflow-instances/${instanceId}/history`),
      {
        headers: this.getAuthHeaders(),
      },
    );

    if (!response.ok) {
      throw new Error("Failed to fetch workflow history");
    }

    return response.json();
  }

  async getAvailableActions(instanceId: number): Promise<any> {
    const response = await fetch(
      getApiUrl(`/api/v1/workflow-instances/${instanceId}/available-actions`),
      {
        headers: this.getAuthHeaders(),
      },
    );

    if (!response.ok) {
      throw new Error("Failed to fetch available actions");
    }

    return response.json();
  }
}

export const workflowAPI = new WorkflowAPI();
